staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go

```diff
@@ -42,24 +52,94 @@ const (
+// AggregationController periodically check for changes in OpenAPI specs of APIServices and update/remove them if necessary.
 type AggregationController struct {
-	openAPIAggregationManager aggregator.SpecAggregator
-	queue                     workqueue.RateLimitingInterface
-	downloader                *aggregator.Downloader
+	aggregator aggregator.SpecAggregator
+	queue      workqueue.RateLimitingInterface
+	downloader *download.Downloader
+
+	lock     sync.Mutex
+	handlers map[string]http.Handler
 
 	// To allow injection for testing.
 	syncHandler func(key string) (syncAction, error)
 }
 
+// NewAggregationControllerWithLocalHandlers creates a new OpenAPI aggregation controller with local delegate fake
+// handlers in addition.
+func NewAggregationControllerWithLocalHandlers(downloader *download.Downloader, a aggregator.SpecAggregator, aggregatorSpec *spec.Swagger, delegationTarget server.DelegationTarget) *AggregationControllerWithLocalHandlers {
+	// register aggregator spec
+	i := 0
+	name, service := localFakeService(i)
+	utilruntime.Must(a.AddUpdateService(name, service))
+	utilruntime.Must(a.UpdateSpec(name, aggregatorSpec, ""))
+	i++
+
+	c := &AggregationControllerWithLocalHandlers{
+		AggregationController: NewAggregationController(downloader, a),
+		delegationTarget:      delegationTarget,
+	}
+
+	// register all other specs and handlers
+	for delegate := delegationTarget; delegate != nil; delegate = delegate.NextDelegate() {
+		handler := delegate.UnprotectedHandler()
+		if handler == nil {
+			continue
+		}
+		name, service := localFakeService(i)
+		a.AddUpdateService(name, service)
+		c.AggregationController.handlers[name] = handler
+		// the spec update will come through periodic queuing through the Run method
+
+		i++
+	}
+
+	return c
+}
+
+type AggregationControllerWithLocalHandlers struct {
+	*AggregationController
+
+	delegationTarget server.DelegationTarget
+}
+
+// localFakeServiceName return a name to be used for the i'th local delegate handler.
+func localFakeServiceName(i int) string {
+	return fmt.Sprintf(localDelegateChainNamePattern, i, "v1")
+}
+
+// localFakeService creates a fake APIService to be used for the i'th local delegate handler.
+// TODO: add knowledge about which APIService belongs to which delegate and register each delegate's with the correct APIServices
+func localFakeService(i int) (string, *apiregistration.APIService) {
+	name := localFakeServiceName(i)
+	return name, &apiregistration.APIService{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+		Spec:       apiregistration.APIServiceSpec{Group: name, Version: "v1"},
+	}
+}
+
+func (c *AggregationControllerWithLocalHandlers) Run(stopCh <-chan struct{}) {
+	// resync local specs every second
+	go wait.Until(func() {
+		i := 1 // skip aggregator
+		for delegate := c.delegationTarget; delegate != nil; delegate = delegate.NextDelegate() {
+			name, _ := localFakeService(i)
+			c.queue.AddAfter(name, time.Second)
+			i++
+		}
+	}, time.Second, stopCh)
```

Comment:
for each key resynced successfully, the queue [forgets the key and adds it back](https://github.com/kubernetes/kubernetes/blob/94a1ad46562268084c174d8af874330b3bfd15e3/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L178-L188) with one minute delay

depending on how the queue is implemented, I feel here we are racing and/or flooding keys

Reply:
Good point. Fixed.